examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,363 @@
-)
-op.add_option(
-    "--n-features",
-    type=int,
-    default=10000,
-    help="Maximum number of features (dimensions) to extract from text.",
-)
-op.add_option(
-    "--verbose",
-    action="store_true",
-    dest="verbose",
-    default=False,
-    help="Print progress reports inside k-means algorithm.",
+dataset = fetch_20newsgroups(
+    remove=("headers", "footers", "quotes"),
+    subset="all",
+    categories=categories,
+    shuffle=True,
+    random_state=42,
 )
 
-print(__doc__)
+labels = dataset.target
+true_k = np.unique(labels).shape[0]
+
+print(f"{len(dataset.data)} documents - {true_k} categories")
 
+# %%
+# Comparing Clustering
+# ====================
+#
+# Here we define a function to score different clustering pipelines using
+# several metrics.
+#
+# Given the knowledge of the ground truth class assignments of the samples, it
+# is possible to define metrics to test for different properties of a cluster:
+#
+# - homogeneity, which quantifies cluster containing only members of a single
+#   class;
+#
+# - completeness, which quantifies cluster having members of a given class are
+#   assigned to the same cluster;
+#
+# - V-measure, the harmonic mean of completeness and homogeneity;
+#
+# - Rand-Index, which measures the similarity between pairs of clusters;
+#
+# - Adjusted Rand-Index, a change-adjusted Rand-Index.
+#
+# If the ground truth labels are not known, evaluation can only be performed
+# using the model results itself. In that case, the Silhouette Coefficient
+# comes in handy.
+#
+# For more reference, see :ref:`clustering_evaluation`.
+
+from collections import defaultdict
+from sklearn import metrics
+from time import time
 
-def is_interactive():
-    return not hasattr(sys.modules["__main__"], "__file__")
+scores = defaultdict(list)
+train_times = defaultdict(list)
 
 
-if not is_interactive():
-    op.print_help()
-    print()
+def fit_and_evaluate(km, X, name=None):
+    name = km.__class__.__name__ if name is None else name
 
-# work-around for Jupyter notebook and IPython console
-argv = [] if is_interactive() else sys.argv[1:]
-(opts, args) = op.parse_args(argv)
-if len(args) > 0:
-    op.error("this script takes no arguments.")
-    sys.exit(1)
+    t0 = time()
+    km.fit(X)
+    train_time = time() - t0
+
+    # store the timing for each fit
+    train_times["estimator"].append(name)
+    train_times["train time"].append(train_time)
+
+    # store the scores of the clustering
+    scores["estimator"].append(name)
+    scores["Homogeneity"].append(metrics.homogeneity_score(labels, km.labels_))
+    scores["Completeness"].append(metrics.completeness_score(labels, km.labels_))
+    scores["V-measure"].append(metrics.v_measure_score(labels, km.labels_))
+    scores["Adjusted Rand-Index"].append(
+        metrics.adjusted_rand_score(labels, km.labels_)
+    )
+    scores["Silhouette Coefficient"].append(
+        metrics.silhouette_score(X, km.labels_, sample_size=2000)
+    )
```

Comment:
I think we should instead collect a `score` dict for each run and then `pprint(score)` (`from pprint import pprint`) to have each call of `fit_and_evaluate` print the evaluation of that call to `stdout` below each cell while accumulating all the score dicts in a list of dicts that is then fed to `pandas.DataFrame` for plotting.

Reply:
https://stackoverflow.com/a/53831756/163740